strider/query_planner.py

```diff
@@ -209,3 +210,78 @@ def get_query_graph_edge_kps(
                 "reverse": og_edge["edge_reverse"],
             })
     return kps
+
+
+def get_next_qedge(qgraph):
+    """Get next qedge to solve."""
+    qgraph = copy.deepcopy(qgraph)
+    for qnode in qgraph["nodes"].values():
+        if qnode.get("ids") is not None:
+            qnode["ids"] = len(qnode["ids"])
+        else:
+            qnode["ids"] = 1_000_000
+    pinnednesses = {
+        qnode_id: get_pinnedness(qgraph, qnode_id)
+        for qnode_id in qgraph["nodes"]
+    }
+    efforts = {
+        qedge_id: math.log(
+            qgraph["nodes"][qedge["subject"]]["ids"]
+        ) + math.log(
+            qgraph["nodes"][qedge["object"]]["ids"]
+        )
+        for qedge_id, qedge in qgraph["edges"].items()
+    }
+    edge_priorities = {
+        qedge_id: pinnednesses[qedge["subject"]] + pinnednesses[qedge["object"]] - efforts[qedge_id]
+        for qedge_id, qedge in qgraph["edges"].items()
+    }
+    qedge_id = max(edge_priorities, key=edge_priorities.get)
+    return qedge_id, qgraph["edges"][qedge_id]
+
+
+N = 1000000  # total number of nodes
+R = 25  # number of edges per node
+
+
+def get_pinnedness(qgraph, qnode_id):
+    """Get pinnedness of each node."""
+    adjacency_mat = get_adjacency_matrix(qgraph)
+    return -_get_pinnedness_recursively(
```

Comment:
Why is there a minus sign here? I think it's easy to miss and should be more obvious.

Reply:
I modified some naming/documentation to try and make this more readable.